src/tools/miri/tests/pass/float.rs

```diff
@@ -151,6 +170,73 @@ fn basic() {
 fn casts() {
+    /* f16 -> int */
+
+    // f16 -> i8
+    test_both_cast::<f16, i8>(127.99, 127);
+    test_both_cast::<f16, i8>(-128.99, -128);
+
+    // f16 -> i16
+    test_both_cast::<f16, i16>(0.0, 0);
+    test_both_cast::<f16, i16>(-0.0, 0);
+    test_both_cast::<f16, i16>(/*0x1p-149*/ f16::from_bits(0x0001), 0);
+    test_both_cast::<f16, i16>(/*-0x1p-149*/ f16::from_bits(0x8001), 0);
+    test_both_cast::<f16, i16>(/*0x1.19999ap+0*/ f16::from_bits(0x3f8c), 1);
+    test_both_cast::<f16, i16>(/*-0x1.19999ap+0*/ f16::from_bits(0xbf8d), -1);
+    test_both_cast::<f16, i16>(1.9, 1);
+    test_both_cast::<f16, i16>(-1.9, -1);
+    test_both_cast::<f16, i16>(5.0, 5);
+    test_both_cast::<f16, i16>(-5.0, -5);
+    test_both_cast::<f16, i16>(32767.0, i16::MAX);
+    test_both_cast::<f16, i16>(-32768.0, i16::MIN);
+    // unrepresentable casts
+    assert_eq::<i16>(32767.0f16 as i16, i16::MAX);
+    assert_eq::<i16>(-32768.0f16 as i16, i16::MIN);
+    assert_eq::<i16>(32780.0f16 as i16, i16::MAX);
+    assert_eq::<i16>(-32780.0f16 as i16, i16::MIN);
+    // FIXME(f16_f128): enable once constants are available
+    // assert_eq::<i16>(f16::INFINITY as i16, i16::MAX);
+    // assert_eq::<i16>(f16::NEG_INFINITY as i16, i16::MIN);
+    // assert_eq::<i16>(f16::NAN as i16, 0);
+    // assert_eq::<i16>((-f16::NAN) as i16, 0);
+
+    // f16 -> u16
+    test_both_cast::<f16, u16>(0.0, 0);
+    test_both_cast::<f16, u16>(-0.0, 0);
+    test_both_cast::<f16, u16>(-0.9999999, 0);
+    test_both_cast::<f16, u16>(/*0x1p-149*/ f16::from_bits(0x1), 0);
+    test_both_cast::<f16, u16>(/*-0x1p-149*/ f16::from_bits(0x8001), 0);
+    test_both_cast::<f16, u16>(/*0x1.19999ap+0*/ f16::from_bits(0x3f8d), 1);
+    test_both_cast::<f16, u16>(1.9, 1);
+    test_both_cast::<f16, u16>(5.0, 5);
+    test_both_cast::<f16, u16>(32768.0, 0x8000);
+    test_both_cast::<f16, u16>(65408.0, 0u16.wrapping_sub(128));
+    test_both_cast::<f16, u16>(/*-0x1.ccccccp-1*/ f16::from_bits(0xbb33), 0);
+    test_both_cast::<f16, u16>(/*-0x1.fffffep-1*/ f16::from_bits(0xbbff), 0);
+    // TODO
+    // test_both_cast::<f16, u16>((u16::MAX - 16) as f16, u16::MAX - 31); // rounding loss
+    // unrepresentable casts
+    assert_eq::<u16>((u16::MAX - 8) as f16 as u16, u16::MAX); // rounds up and then becomes unrepresentable
+    assert_eq::<u16>(65536.0f16 as u16, u16::MAX);
+    assert_eq::<u16>(-5.0f16 as u16, 0);
+    assert_eq::<u16>(f16::MAX as u16, 65504);
+    assert_eq::<u16>(f16::MIN as u16, 0);
+    // FIXME(f16_f128): enable once constants are available
+    // assert_eq::<u16>(f16::INFINITY as u16, u16::MAX);
+    // assert_eq::<u16>(f16::NEG_INFINITY as u16, 0);
+    // assert_eq::<u16>(f16::NAN as u16, 0);
+    // assert_eq::<u16>((-f16::NAN) as u16, 0);
+
+    // f16 -> i32
+    assert_eq::<i32>(f16::MAX as i32, 65504);
+    assert_eq::<i32>(f16::MIN as i32, -65504);
+
+    // f16 -> u32
+    assert_eq::<u32>(f16::MAX as u32, 65504);
+    assert_eq::<u32>(f16::MIN as u32, 0);
```

Comment:
Does this match rustc tests, or is rustc not ready yet to run all of these?
I can't really check whether your values make sense; so far I always did this by just running the test with rustc and then making sure Miri behavior matches rustc behavior.

Reply:
rustc should be able to run the `f16`/`f128` on at least Linux x86 and aarch64, but many other platforms are missing the f16/f128 symbols until compiler builtins gets updated. Not sure if you meant fully testing in CI or just locally.

However, I still don't have the constants needed to fully test casting on the new types. So I moved only the test refactoring to its own PR (https://github.com/rust-lang/rust/pull/126727), and am just going to close this one until https://github.com/rust-lang/rust/pull/126608 can land.

Reply:
Also there are some ABI issues on x86 without SSE2, which might explain the failure at https://github.com/rust-lang/rust/actions/runs/9589520480/job/26443432727?pr=126718#step:25:3379 (haven't looked too much into it)